File: pbmx_ffi/java/io/rmf/pbmx/Block.java
package io.rmf.pbmx;

import io.rmf.pbmx.ffi.LibPbmx;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.WritableByteChannel;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.LongByReference;

public final class Block {
    Block(Pointer handle) {
        this.handle = handle;
    }

    public Id id() {
        return new Id(LibPbmx.INSTANCE.pbmx_block_id(this.handle));
    }

    public Fingerprint signer() {
        return new Fingerprint(LibPbmx.INSTANCE.pbmx_block_id(this.handle));
    }

    public static Block readFrom(ByteBuffer buf) {
        Pointer handle = LibPbmx.INSTANCE.pbmx_import_block(buf, buf.remaining());
        return new Block(handle);
    }

    public void writeTo(WritableByteChannel channel) throws IOException {
        LongByReference len = new LongByReference();
        int r = LibPbmx.INSTANCE.pbmx_export_block(this.handle, null, len);
        assert r != 0;

        ByteBuffer buf = ByteBuffer.allocate((int)len.getValue());
        r = LibPbmx.INSTANCE.pbmx_export_block(this.handle, buf, len);
        assert r != 0;

        channel.write(buf);
    }

    @Override
    protected void finalize() {
        LibPbmx.INSTANCE.pbmx_delete_block(this.handle);
    }

    Pointer handle;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/Block.kt
package io.rmf.pbmx

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.payloads.Payload
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.channels.WritableByteChannel

class Block internal constructor(internal var handle: Pointer) {

    val id get() = Id(FFI.pbmx_block_id(this.handle))

    val signer get() = Fingerprint(FFI.pbmx_block_signer(this.handle))

    val payloads: List<Payload>
        get() {
            val length = LongByReference()
            var r = FFI.pbmx_payloads(this.handle, null, length)
            assert(r == 0)

            val payloads = Array(length.value.toInt()) { Pointer.NULL }
            r = FFI.pbmx_payloads(this.handle, payloads, length)
            assert(r != 0)

            return payloads.map { Payload.from(it) }.toList()
        }

    @Throws(IOException::class)
    fun writeTo(channel: WritableByteChannel) {
        val len = LongByReference()
        var r = FFI.pbmx_export_block(this.handle, null, len)
        assert(r == 0)

        val buf = ByteBuffer.allocate(len.value.toInt())
        r = FFI.pbmx_export_block(this.handle, buf, len)
        assert(r != 0)

        channel.write(buf)
    }

    protected fun finalize() {
        FFI.pbmx_delete_block(this.handle)
    }

    companion object {

        fun readFrom(buf: ByteBuffer): Block {
            val handle = FFI.pbmx_import_block(buf, buf.remaining().toLong())
            return Block(handle)
        }
    }
}
